src/NuGet.Core/NuGet.Configuration/PackageNamespaces/SearchTree.cs

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace NuGet.Configuration
+{
+    internal class SearchTree
+    {
+        private readonly SearchNode _root;
+
+        internal SearchTree(PackageNamespacesConfiguration configuration)
+        {
+            if (configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
+            _root = new SearchNode();
+
+            foreach (string packageSourceKey in configuration.Namespaces.Keys)
+            {
+                foreach (string nugetNamespace in configuration.Namespaces[packageSourceKey])
+                {
+                    Add(packageSourceKey, nugetNamespace);
+                }
+            }
+        }
+
+        private void Add(string packageSourceKey, string namespaceId)
+        {
+            SearchNode currentNode = _root;
+
+            if (namespaceId == null)
+            {
+                throw new ArgumentNullException(nameof(namespaceId));
+            }
+
+            // To prevent from unwanted behaviour.
+            if (namespaceId.Length > PackageNamespacesConfiguration.PackageIdMaxLength)
+            {
+                throw new ArgumentOutOfRangeException(nameof(namespaceId));
+            }
+
+#pragma warning disable CA1308 // Normalize strings to uppercase
+            packageSourceKey = packageSourceKey.ToLowerInvariant().Trim();
+            namespaceId = namespaceId.ToLowerInvariant().Trim();
+#pragma warning restore CA1308 // Normalize strings to uppercase
+
+            for (int i = 0; i < namespaceId.Length; i++)
+            {
+                char c = namespaceId[i];
+
+                if (c == '*')
```

Comment:
If user defines an invalid namespace such as `NuGet.*Common` then as per the above logic, we define a namespace on `NuGet.*` which may be incorrect. I think we should throw an exception because the configuration of `NuGet.*Common` is currently invalid. One way to do it is check the first index of `*` in the namespace, `if first_index != namespaceId.Length-1` then throw an exception.

Reply:
I think we should validate the namespaces, early when we process the configs, that way we only do it once.

Reply:
That being said, we can just create a follow up, if there already isn't one to get that done.

Reply:
I created https://github.com/NuGet/Home/issues/10941 to keep track of this concern. It's more of `PackageNamespacesConfiguration` validation issue than problem I'm trying to solve.